keps/sig-api-machinery/20190129-kflag.md

```diff
@@ -0,0 +1,205 @@
+  config) is a potential decision point, prior to which flag precedence must be enforced.
+* Components should be able to apply the parsed flag values to structs that only represent a subset
+  of all flag values. For instance, if a component has a Flags struct and a Config struct,
+  it may need to apply values to each independently.
+* Components must be able to specify custom behavior for merging values during application.
+  Feature gates, for example, are merged piecemeal between the command-line and config files, with
+  gates specified on the command-line taking precedence.
+* Application of flag values should only consider the flags specified on the command line. It should
+  be careful, for example, to not overwrite values loaded from config files with the default values
+  of omitted flags.
+
+
+### Implementation Details
+
+Fundamentally, separating parsing from application requires that we have a scratch space to parse
+the initial values into, and a way to map that scratch space onto target objects.
+
+To maintain type safety, this proposal recommends a two-layered approach, where kflag handles the
+common types, and components handle the component-specific types:
+* kflag Implements:
+  * Statically typed flag registration helpers that allocate scratch space.
+  * Statically typed helpers to apply the values of parsed flags to arbitrary targets.
+* Components Implement:
+  * Aggregate flag registration functions, which use kflag helpers to register groups of flags the
+    component cares about.
+  * Aggregate flag application functions, which use kflag helpers to apply a subset of flag values
+    to an arbitrary structure determined by the component.
+
+This proposal recommends the approach in this example PR:
+* https://github.com/kubernetes/kubernetes/pull/73494
+
+See also @sttts's prototype for an alternative approach that predates this KEP:
+* https://github.com/kubernetes/kubernetes/pull/72037
+
+#### Experimental First
+
+Kflag will initially be implemented in `k8s.io/component-base/experimental/cli/kflag`, and can
+be graduated to `k8s.io/kflag` (external repo) when we are satisfied with the progress.
+
+During the implementation, we will maintain an open PR with an example conversion of an existing
+component, to demonstrate kflag and help us determine when it is ready to graduate from experimental.
+
+### Risks and Mitigations
+
+#### Maintenance
+The kflag package is a wrapper around an existing package, pflag.
+The pflag package has a stable API, so we would at most expect to extend kflag when pflag is
+extended, but only in the event that we really need that extension. We expect this to be infrequent.
+Given that kflag should make components easier to maintain (it reduces the complexity of the
+bootstrap process by making the flag precedence implementation easier to use and reason about),
+the maintenance of kflag itself is likely time well spent.
+
+#### Migrating existing components
```

Comment:
Have you considered just building something directly for a component, and then generalizing it after that? The risk I see in this KEP is that it may be building something overly general.